pkg/gadgets/snapshot/socket/tracer/tracer.go

```diff
@@ -193,3 +200,69 @@ func RunCollector(pid uint32, podname, namespace, node string, proto socketcolle
 
 	return sockets, nil
 }
+
+// ---
+
+type Tracer struct {
+	// visitedNamespaces is a map where the key is the netns inode number and
+	// the value is the pid of one of the containers that share that netns. Such
+	// pid is used by NetnsEnter. TODO: Improve NetnsEnter to also work with the
+	// netns directly.
+	visitedNamespaces map[uint64]uint32
+	protocols         string
+	eventHandler      func([]*socketcollectortypes.Event)
+}
+
+func (g *GadgetDesc) NewInstance() (gadgets.Gadget, error) {
+	tracer := &Tracer{
+		visitedNamespaces: make(map[uint64]uint32),
+	}
+	return tracer, nil
+}
+
+func (t *Tracer) Init(gadgetCtx gadgets.GadgetContext) error {
+	params := gadgetCtx.GadgetParams()
+	t.protocols = params.Get(ParamProto).AsString()
+	return nil
+}
+
+func (t *Tracer) AttachContainer(container *containercollection.Container) error {
+	if container.Pid == 0 {
+		panic("Pid is 0 for container " + container.Name + " (" + container.ID + ")")
+	}
```

Comment:
Some comments about this:
- I don't think we need this kind of check here. All the previous logic should guarantee that a container has its PID set. 
- We're not checking this on any other gadget.
- Panicking is too extreme: we don't want to cause the gadget pod to restart because a container is missing its PID.

Reply:
I'm fine with both solutions but I have a question: Should we consider that people could use the trace directly and may pass a container with 0 pid? 

> Panicking is too extreme: we don't want to cause the gadget pod to restart because a container is missing its PID.
>
My reason for panicking is exactly that all the logic in IG for collecting the containers is made to never miss its PID. If it happens is because we have an important bug because we rely on that PID for most things. Maybe panicking is too much but at least the whole gadget run should fail.

Reply:
> Should we consider that people could use the trace directly and may pass a container with 0 pid?

That's a good question, probably people could use this gadget alone but IMO there is still some work on the API to make it easy to use, currently the call will be something like:

```golang

t := NewInstance(...)

t.AttachContainer(...)
t.AttachContainer(...)

t.Run()
```

> Maybe panicking is too much but at least the whole gadget run should fail.

I think it will fail while calling `containerutils.GetNetNs(int(pid))`, right?

Reply:
Yep, it'll fail there. OK, we can skip that check and come back to it when developing an example using the tracer directly 🙂.